src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs

```diff
@@ -0,0 +1,322 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Arcus.Messaging.Abstractions;
+using Arcus.Messaging.Pumps.Abstractions;
+using GuardNet;
+using Microsoft.Azure.ServiceBus;
+using Microsoft.Azure.ServiceBus.Core;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Arcus.Messaging.Pumps.ServiceBus
+{
+    /// <summary>
+    ///     Message pump for processing messages on an Azure Service Bus entity
+    /// </summary>
+    /// <typeparam name="TMessage">Type of expected message payload</typeparam>
+    public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessage, AzureServiceBusMessageContext>
+    {
+        private bool _isHostShuttingDown;
+        private MessageReceiver _messageReceiver;
+        private readonly MessageHandlerOptions _messageHandlerOptions;
+
+        /// <summary>
+        ///     Constructor
+        /// </summary>
+        /// <param name="configuration">Configuration of the application</param>
+        /// <param name="serviceProvider">Collection of services that are configured</param>
+        /// <param name="logger">Logger to write telemetry to</param>
+        protected AzureServiceBusMessagePump(IConfiguration configuration, IServiceProvider serviceProvider,
```

Comment:
to me, it's a code smell if you have to inject the serviceprovider in a ctor, effectively making your code dependant on an external mechanism for resolving dependencies. Is there a reason why the dependency isn't injected?

Reply:
It was not on purpose so I'll refactor it.

Reply:
Now I remember.

The reason for this is that every implementation you create, such as `OrdersMessagePump` in our tests, should not have to worry about the internal plumbing.

We can remove it but then all the internal dependencies for `AzureServiceBusMessagePump`, and other sinks in the future, will be leaked which makes them less portable.

That's why I'd stick to this approach, frankly, unless there is a strong argument against it.

So bottom line - Yes I agree, but here I see it as an exception to the rule.

Reply:
Example without removal:
```csharp
public class OrdersMessagePump : AzureServiceBusMessagePump<Order>
    {
        private readonly IEventGridPublisher _eventGridPublisher;

        public OrdersMessagePump(IConfiguration configuration, IServiceProvider serviceProvider, ILogger<OrdersMessagePump> logger)
            : base(configuration, serviceProvider, logger)
        {
            var eventGridTopic = configuration.GetValue<string>("EVENTGRID_TOPIC_URI");
            var eventGridKey = configuration.GetValue<string>("EVENTGRID_AUTH_KEY");

            _eventGridPublisher = EventGridPublisherBuilder
                .ForTopic(eventGridTopic)
                .UsingAuthenticationKey(eventGridKey)
                .Build();
        }

        protected override async Task ProcessMessageAsync(Order orderMessage, AzureServiceBusMessageContext messageContext, MessageCorrelationInfo correlationInfo, CancellationToken cancellationToken)
        {
            // Logic
        }
    }
```

Example with DI according to the text books:
```csharp
public class OrdersMessagePump : AzureServiceBusMessagePump<Order>
    {
        private readonly IEventGridPublisher _eventGridPublisher;

        public OrdersMessagePump(IConfiguration configuration, AzureServiceBusMessagePumpSettings messagePumpSettings, ILogger<OrdersMessagePump> logger)
            : base(configuration, messagePumpSettings, logger)
        {
            var eventGridTopic = configuration.GetValue<string>("EVENTGRID_TOPIC_URI");
            var eventGridKey = configuration.GetValue<string>("EVENTGRID_AUTH_KEY");

            _eventGridPublisher = EventGridPublisherBuilder
                .ForTopic(eventGridTopic)
                .UsingAuthenticationKey(eventGridKey)
                .Build();
        }

        protected override async Task ProcessMessageAsync(Order orderMessage, AzureServiceBusMessageContext messageContext, MessageCorrelationInfo correlationInfo, CancellationToken cancellationToken)
        {
            // Logic
        }
    }
```

Reply:
Adding @stijnmoreels to get his input as well.

Reply:
Yes, when implementating a custom message pump myself; I too came across this setup. I thought that it was to provide some "stable" ctor signature with some "major" app dependencies so we don't have to worry (too) much about future signature changes.
So, yes; I agree with both 🙃 . Code smell but probably a good, stable setup.